File: python/919. Complete Binary Tree Inserter.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class CBTInserter(object):

    def __init__(self, root):
        """
        :type root: TreeNode
        """
        self.root = root
        self.nodes = []
        que = [root]
        while que:
            root = que.pop(0)
            self.nodes.append(root)
            if root.left:
                que.append(root.left)
            if root.right:
                que.append(root.right)
        

    def insert(self, v):
        """
        :type v: int
        :rtype: int
        """
        self.nodes.append(TreeNode(v))
        n = len(self.nodes)
        idx = 0
        if n&1:
            idx = (n-1)/2 - 1
            self.nodes[idx].right = self.nodes[-1]
        else:
            idx = n/2 - 1
            self.nodes[idx].left = self.nodes[-1]
        return self.nodes[idx].val

    def get_root(self):
        """
        :rtype: TreeNode
        """
        return self.root
        


# Your CBTInserter object will be instantiated and called as such:
# obj = CBTInserter(root)
# param_1 = obj.insert(v)
# param_2 = obj.get_root()

File: go/src/leetcode/919/919. Complete Binary Tree Inserter.go
/**
@author ZhengHao Lou
Date    2022/07/25
*/
package main

import "fmt"

/**
https://leetcode.cn/problems/complete-binary-tree-inserter/
*/
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type CBTInserter struct {
	nodes []*TreeNode
	root  *TreeNode
}

func Constructor(root *TreeNode) CBTInserter {
	var tree = CBTInserter{
		nodes: []*TreeNode{{}},
		root:  root,
	}
	if root == nil {
		return tree
	}
	var que = []*TreeNode{root}
	for len(que) != 0 {
		node := que[0]
		que = que[1:]
		if node.Left != nil {
			que = append(que, node.Left)
		}
		if node.Right != nil {
			que = append(que, node.Right)
		}
		tree.nodes = append(tree.nodes, node)
	}
	return tree
}

func (this *CBTInserter) Insert(val int) int {
	node := &TreeNode{Val: val}
	this.nodes = append(this.nodes, node)
	index := len(this.nodes) - 1
	var parent = this.nodes[index>>1]
	if index&1 == 0 {
		parent.Left = node
	} else {
		parent.Right = node
	}
	return parent.Val
}

func (this *CBTInserter) Get_root() *TreeNode {
	return this.nodes[1]
}

func main() {
	root := &TreeNode{
		Val:  1,
		Left: &TreeNode{Val: 2},
	}
	obj := Constructor(root)
	obj.Insert(3)
	obj.Insert(4)
	fmt.Println(obj.Get_root())
}

/**
 * Your CBTInserter object will be instantiated and called as such:
 * obj := Constructor(root);
 * param_1 := obj.Insert(val);
 * param_2 := obj.Get_root();
 */
